stan/math/prim/fun/quantile.hpp

```diff
@@ -0,0 +1,119 @@
+
+/**
+ * Return sample quantiles corresponding to the given probabilities.
+ * The smallest observation corresponds to a probability of 0 and the largest to
+ * a probability of 1.
+ *
+ * Implements algorithm 7 from Hyndman, R. J. and Fan, Y.,
+ * Sample quantiles in Statistical Packages (R's default quantile function)
+ *
+ * @tparam T An Eigen type with either fixed rows or columns at compile time, or
+ * std::vector<double>
+ * @param samples_vec Numeric vector whose sample quantiles are wanted
+ * @param p Probability with value between 0 and 1.
+ * @return Sample quantile.
+ * @throw std::invalid_argument If any element of samples_vec is NaN, or size 0.
+ * @throw std::domain_error If `p<0` or `p>1`.
+ */
+template <typename T, require_vector_t<T>* = nullptr,
+          require_vector_vt<std::is_arithmetic, T>* = nullptr>
+inline double quantile(const T& samples_vec, const double p) {
+  check_not_nan("quantile", "samples_vec", samples_vec);
+  check_nonzero_size("quantile", "samples_vec", samples_vec);
+
+  check_not_nan("quantile", "p", p);
+  check_bounded("quantile", "p", p, 0, 1);
+
+  const size_t n_sample = samples_vec.size();
+
+  Eigen::VectorXd x = as_array_or_scalar(samples_vec);
+
+  if (n_sample == 1)
+    return x.coeff(0);
+  else if (p == 0.)
+    return *std::min_element(x.data(), x.data() + n_sample);
+  else if (p == 1.)
+    return *std::max_element(x.data(), x.data() + n_sample);
+
+  double index = (n_sample - 1) * p;
+  size_t lo = std::floor(index);
+  size_t hi = std::ceil(index);
+
+  std::sort(x.data(), x.data() + n_sample, std::less<double>());
+
+  double h = index - lo;
+  return (1 - h) * x.coeff(lo) + h * x.coeff(hi);
+}
+
+/**
+ * Return sample quantiles corresponding to the given probabilities.
+ * The smallest observation corresponds to a probability of 0 and the largest to
+ * a probability of 1.
+ *
+ * Implements algorithm 7 from Hyndman, R. J. and Fan, Y.,
+ * Sample quantiles in Statistical Packages (R's default quantile function)
+ *
+ * @tparam T An Eigen type with either fixed rows or columns at compile time, or
+ * std::vector<double>
+ * @tparam Tp Type of probabilities vector
+ * @param samples_vec Numeric vector whose sample quantiles are wanted
+ * @param ps Vector of probability with value between 0 and 1.
+ * @return Sample quantiles, one for each p in ps.
+ * @throw std::invalid_argument If any of the values are NaN or size 0.
+ * @throw std::domain_error If `p<0` or `p>1` for any p in ps.
+ */
+template <typename T, typename Tp, require_all_vector_t<T, Tp>* = nullptr,
+          require_vector_vt<std::is_arithmetic, T>* = nullptr,
+          require_vector_vt<std::is_arithmetic, Tp>* = nullptr>
+inline std::vector<double> quantile(const T& samples_vec, const Tp& ps) {
+  check_not_nan("quantile", "samples_vec", samples_vec);
+  check_not_nan("quantile", "ps", ps);
+
+  check_nonzero_size("quantile", "samples_vec", samples_vec);
+  check_nonzero_size("quantile", "ps", ps);
```

Comment:
Standard for zero sized vectors is just to return the input vector. Generally we try not to fail on zero sized things. There's cases like in rstanarm where we can have size zero vectors

Reply:
OK, so something like?
```cpp
const size_t n_ps = ps.size();
if (n_ps == 0) 
    return {}
```
Sorry if this is super-basic, learning C++ as I go, here... :-)

Reply:
Yep!

Reply:
And no worries! C++ can be real weird

Reply:
Thanks!

Reply:
Hmm, when I comment out `check_nonzero_size("quantile", "ps", ps);` I fail this test:
```cpp
EXPECT_THROW(quantile(v, std::numeric_limits<Tp>::quiet_NaN()),
               std::invalid_argument);
```
with
```
Expected: quantile(v, std::numeric_limits<Tp>::quiet_NaN()) throws an exception of type std::invalid_argument.
  Actual: it throws nothing.
```
shouldn't `check_not_nan("quantile", "ps", ps);` catch this? Or

Reply:
It was easier to show then tell alone

https://github.com/stan-dev/math/tree/review/quantile-funcs

But essentially [`std::numeric_limits<T>::quiet_NaN()`](https://en.cppreference.com/w/cpp/types/numeric_limits) is only defined for floating point types so you can't pass a vector to it.

Reply:
Thanks! Merged.